File: cn/huohuas001/bot/events/GroupMessageHandler.java
package cn.huohuas001.bot.events;
import cn.huohuas001.bot.HuHoBot;
import cn.huohuas001.bot.events.commands.*;
import cn.huohuas001.bot.state.CommandRepositories;
import cn.huohuas001.huhobotPenguin.spigot.qqbind.AiChat;
import cn.huohuas001.huhobotPenguin.spigot.qqbind.QqBindManager;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import io.github.kloping.qqbot.api.v2.GroupMessageEvent;
import io.github.kloping.qqbot.entities.qqpd.User;
import io.github.kloping.qqbot.entities.qqpd.v2.Contact;
import io.github.kloping.qqbot.impl.ListenerHost;
import io.github.kloping.qqbot.impl.message.v2.BaseMessageEvent;
import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;

public final class GroupMessageHandler extends ListenerHost {
    private final HuHoBot plugin;
    private final CopyOnWriteArrayList<BaseCommand> commands = new CopyOnWriteArrayList<>();
    public GroupMessageHandler(HuHoBot plugin) { this.plugin = plugin; this.commands.add(new PublicCommands()); this.commands.add(new AdministrationCommands()); this.commands.add(new MotdCommands()); this.commands.add(new BindCommands()); }
    public void registerCommand(BaseCommand command) { if (command != null) this.commands.add(command); }
    @ListenerHost.EventReceiver
    public void onGroupMessage(GroupMessageEvent event) {
        if (event == null) return;
        String groupId = event.getGroupOpenId(); if (groupId == null) groupId = event.getGroupId();
        String content = event.getRawMessage().getContent(); if (content == null) return;
        if (!content.contains("查信息")) { if (groupId == null || !isAllowedGroup(groupId)) return; }
        if (dispatchCommand(event)) return;
        if (content.contains("<@") && !content.trim().startsWith("/")) {
            boolean mentioned = isMentionedBot(event);
            if (mentioned) { handleAiChat(event, content, groupId); return; }
        }
        if (groupId != null) forwardFullGroupMessage(groupId, event);
    }
    private boolean isAllowedGroup(String groupId) { List<String> allowed = plugin.getGroupOpenIdList(); return allowed.isEmpty() || allowed.contains(groupId); }
    private boolean dispatchCommand(GroupMessageEvent event) { for (BaseCommand c : commands) { try { if (c.handleMessage(plugin, event)) return true; } catch (Exception e) { plugin.log_error("指令处理异常: " + e.getMessage()); } } return false; }
    private boolean isMentionedBot(GroupMessageEvent event) {
        String botSelfId = null;
        try { io.github.kloping.qqbot.Starter starter = cn.huohuas001.bot.QClient.INSTANCE.getStarter(); if (starter != null && starter.getBot() != null) botSelfId = starter.getBot().getId(); } catch (Throwable ignored) {}
        String botAppId = this.plugin.getBotAppId();
        try { User[] mentions = event.getRawMessage().getMentions(); if (mentions != null) for (int i = 0; i < mentions.length; i++) { User m = mentions[i]; String mid = m.getId(); if (mid != null && botSelfId != null && !botSelfId.isEmpty() && mid.equals(botSelfId)) return true; try { Boolean isBot = m.getBot(); if (isBot != null && isBot) return true; } catch (Throwable ignored) {} if (mid != null && botAppId != null && !botAppId.isEmpty() && mid.equals(botAppId)) return true; } } catch (Throwable ignored) {}
        try { String c = event.getRawMessage().getContent(); if (c != null) { if (botSelfId != null && !botSelfId.isEmpty() && (c.contains("<@!" + botSelfId + ">") || c.contains("<@" + botSelfId + ">"))) return true; if (botAppId != null && !botAppId.isEmpty() && (c.contains("<@!" + botAppId + ">") || c.contains("<@" + botAppId + ">"))) return true; } } catch (Throwable ignored) {}
        return false;
    }
    private void handleAiChat(GroupMessageEvent event, String content, String groupId) {
        try {
            QqBindManager mgr; try { mgr = QqBindManager.getInstance(); } catch (Throwable t) { return; }
            if (!mgr.isAiEnabled()) return;
            String text = content; User[] mentions = event.getRawMessage().getMentions();
            if (mentions != null) for (int i = 0; i < mentions.length; i++) { User m = mentions[i]; String mid = m.getId(); if (mid == null) continue; text = text.replace("<@!" + mid + ">", "").replace("<@" + mid + ">", "").replace("<" + mid + ">", "").replace(mid, ""); }
            text = text.trim(); if (text.isEmpty()) { event.sendMessage("请在 @我 之后输入你想说的话～"); return; }
            String userId = safeUserId(event);
            String reply = AiChat.chat(text, mgr, groupId, userId);
            event.sendMessage(mgr.getQqAiOutputPrefix() + " " + reply);
        } catch (Throwable t) { event.sendMessage("AI 对话失败：" + t.getMessage()); }
    }
    private String safeUserId(GroupMessageEvent event) { try { Contact c = event.getSender(); if (c == null) return "<unknown>"; String oid = c.getOpenid(); if (oid == null) oid = c.getId(); return oid == null ? "<unknown>" : oid; } catch (Throwable t) { return "<unknown>"; } }
    @SuppressWarnings("unchecked")
    private void forwardFullGroupMessage(String groupId, GroupMessageEvent event) {
        try {
            boolean enabled = CommandRepositories.INSTANCE.getGroupSettings().fullForwarding(groupId, plugin.getFullAmount());
            if (!enabled || !plugin.getChatFormat().getPostChat()) return;
            String senderName = "unknown"; Contact contact = event.getSender(); if (contact != null) { String un = contact.getUsername(); if (un != null) senderName = un; }
            BaseMessageEvent baseMsg = event instanceof BaseMessageEvent ? (BaseMessageEvent) event : null;
            JSONObject metadata = baseMsg != null ? baseMsg.getMetadata() : null;
            JSONArray atts = metadata != null ? metadata.getJSONArray("attachments") : null;
            List<String> parts = new ArrayList<>(); String raw = event.getRawMessage().getContent(); String tc = raw != null ? raw.trim() : ""; if (tc.length() > 0) parts.add(tc);
            if (atts != null) { int n = ((Collection<Object>) atts).size(); for (int i = 0; i < n; i++) { JSONObject a = atts.getJSONObject(i); if (a == null) continue; String ct = a.getString("content_type"); if (ct == null) ct = ""; if ("voice".equals(ct)) { String asr = a.getString("asr_refer_text"); parts.add(asr != null && !asr.trim().isEmpty() ? "[语音] [" + asr.trim() + "]" : "[语音]"); } else if (ct.startsWith("image/")) parts.add("[图片]"); else if ("image/gif".equals(ct)) parts.add("[表情包]"); else if (ct.startsWith("video/")) parts.add("[视频]"); else { String fn = a.getString("filename"); parts.add("[文件: " + (fn == null ? "文件" : fn) + "]"); } } }
            if (parts.isEmpty()) return; StringBuilder sb = new StringBuilder(); for (int i = 0; i < parts.size(); i++) { if (i > 0) sb.append(' '); sb.append(parts.get(i)); } String msg = sb.toString();
            User[] mentions = event.getRawMessage().getMentions(); if (mentions != null) for (int i = 0; i < mentions.length; i++) { User m = mentions[i]; String mid = m.getId(), mname = m.getUsername(); if (mid == null || mname == null) continue; msg = msg.replace("<@!" + mid + ">", "@" + mname).replace("<@" + mid + ">", "@" + mname).replace("<" + mid + ">", "@" + mname).replace(mid, "@" + mname); }
            plugin.broadcastMessage(plugin.formatGroupMessage(senderName, plugin.auditText(msg)));
        } catch (Throwable ignored) {}
    }
}

File: common-Bot/src/main/kotlin/cn/huohuas001/bot/events/GroupMessageHandler.kt
package cn.huohuas001.bot.events

import cn.huohuas001.bot.HuHoBot
import cn.huohuas001.bot.QClient
import cn.huohuas001.bot.events.commands.AdministrationCommands
import cn.huohuas001.bot.events.commands.BaseCommand
import cn.huohuas001.bot.events.commands.BindCommands
import cn.huohuas001.bot.events.commands.CheckInCommands
import cn.huohuas001.bot.events.commands.MotdCommands
import cn.huohuas001.bot.events.commands.PublicCommands
import cn.huohuas001.bot.state.CommandRepositories
import cn.huohuas001.huhobotPenguin.spigot.qqbind.AiChat
import cn.huohuas001.huhobotPenguin.spigot.qqbind.QqBindManager
import com.alibaba.fastjson.JSONArray
import com.alibaba.fastjson.JSONObject
import io.github.kloping.qqbot.api.v2.GroupMessageEvent
import io.github.kloping.qqbot.entities.qqpd.User
import io.github.kloping.qqbot.entities.qqpd.v2.Contact
import io.github.kloping.qqbot.impl.ListenerHost
import io.github.kloping.qqbot.impl.message.v2.BaseMessageEvent
import java.util.concurrent.CopyOnWriteArrayList

/**
 * QQ 群消息入口：命令分发 → @机器人 AI 对话 → 全量转发。
 */
class GroupMessageHandler(private val plugin: HuHoBot) : ListenerHost() {

    private val commands: CopyOnWriteArrayList<BaseCommand> = CopyOnWriteArrayList()

    init {
        commands.add(PublicCommands())
        commands.add(AdministrationCommands())
        commands.add(MotdCommands())
        commands.add(BindCommands())
        commands.add(CheckInCommands())
    }

    /** 注册额外的群命令处理器。 */
    fun registerCommand(command: BaseCommand?) {
        if (command != null) commands.add(command)
    }

    @ListenerHost.EventReceiver
    fun onGroupMessage(event: GroupMessageEvent) {
        if (event == null) return
        val groupId: String? = event.groupOpenId ?: event.groupId
        val content = event.rawMessage?.content ?: return

        // 查询open ID（OpenId 查询）与查信息（统计卡片）命令始终放行（任何群可用）；
        // 其余命令仅允许配置的群。1.5.2：两命令更名（原 查信息/个人信息）。
        // “查询open” 用前缀匹配兼容大小写/空格变体（查询OpenID、查询openid 等）。
        if (!content.contains("查信息") && !content.contains("查询open")) {
            if (groupId == null || !isAllowedGroup(groupId)) return
        }

        // 1. 依次尝试命令分发
        if (dispatchCommand(event)) return

        // 2. @机器人触发 AI 对话（非 / 开头）
        if (content.contains("<@") && !content.trim().startsWith("/")) {
            if (isMentionedBot(event)) {
                handleAiChat(event, content, groupId)
                return
            }
        }

        // 3. 全量消息转发到游戏
        if (groupId != null) forwardFullGroupMessage(groupId, event)
    }

    /** 是否为允许交互的群（空配置表示不限制）。 */
    private fun isAllowedGroup(groupId: String): Boolean {
        val allowed = plugin.groupOpenIdList()
        return allowed.isEmpty() || groupId in allowed
    }

    /** 遍历命令处理器尝试处理消息。 */
    private fun dispatchCommand(event: GroupMessageEvent): Boolean {
        for (command in commands) {
            try {
                if (command.handleMessage(plugin, event)) return true
            } catch (e: Exception) {
                plugin.log_error("指令处理异常: ${e.message}")
            }
        }
        return false
    }

    /** 判断消息是否 @ 了机器人。 */
    private fun isMentionedBot(event: GroupMessageEvent): Boolean {
        var botSelfId: String? = null
        try {
            val starter = QClient.starter
            if (starter?.bot != null) botSelfId = starter.bot.id
        } catch (_: Throwable) {
        }
        val botAppId = plugin.botAppId
        // 1) mentions 数组中含机器人
        try {
            val mentions: Array<User>? = event.rawMessage?.mentions
            if (mentions != null) {
                for (member in mentions) {
                    val memberId = member.id ?: continue
                    if (botSelfId != null && botSelfId.isNotEmpty() && memberId == botSelfId) return true
                    try {
                        if (member.bot == true) return true
                    } catch (_: Throwable) {
                    }
                    if (botAppId != null && botAppId.isNotEmpty() && memberId == botAppId) return true
                }
            }
        } catch (_: Throwable) {
        }
        // 2) 文本中的 @ 片段
        try {
            val text = event.rawMessage?.content
            if (text != null) {
                if (botSelfId != null && botSelfId.isNotEmpty() &&
                    (text.contains("<@!$botSelfId>") || text.contains("<@$botSelfId>"))
                ) return true
                if (botAppId != null && botAppId.isNotEmpty() &&
                    (text.contains("<@!$botAppId>") || text.contains("<@$botAppId>"))
                ) return true
            }
        } catch (_: Throwable) {
        }
        return false
    }

    /** 处理 @机器人 的 AI 对话。 */
    private fun handleAiChat(event: GroupMessageEvent, content: String, groupId: String?) {
        try {
            val manager = try {
                QqBindManager.getInstance()
            } catch (_: Throwable) {
                return
            }
            if (!manager.isAiEnabled) return

            // 去除 @ 片段后取正文
            var text = content
            val mentions: Array<User>? = event.rawMessage?.mentions
            if (mentions != null) {
                for (member in mentions) {
                    val memberId = member.id ?: continue
                    text = text.replace("<@!$memberId>", "")
                        .replace("<@$memberId>", "")
                        .replace("<$memberId>", "")
                        .replace(memberId, "")
                }
            }
            text = text.trim()
            if (text.isEmpty()) {
                event.sendMessage("请在 @我 之后输入你想说的话～")
                return
            }
            val userId = safeUserId(event)
            val reply = AiChat.chat(text, manager, groupId, userId)
            val aiContent = "${manager.qqAiOutputPrefix} $reply"
            // 0.1.5.2：AI 对话日志仅写入文件，不再上控制台（隐私泄露修复）
            QqBindManager.logVerbose("[AI对话] 回复内容长度=${aiContent.length} 准备发送")
            try {
                event.sendMessage(aiContent)
            } catch (replyError: Throwable) {
                QqBindManager.logVerbose("[AI对话] sendMessage 失败: ${replyError.message}")
            }
        } catch (t: Throwable) {
            event.sendMessage("AI 对话失败：${t.message}")
        }
    }

    /** 安全提取发送者 OpenId。 */
    private fun safeUserId(event: GroupMessageEvent): String = try {
        val contact: Contact? = event.sender
        val openid = contact?.openid ?: contact?.id
        openid ?: "<unknown>"
    } catch (_: Throwable) {
        "<unknown>"
    }

    /** 全量转发群消息到游戏（含图片/语音/文件占位）。 */
    @Suppress("UNCHECKED_CAST")
    private fun forwardFullGroupMessage(groupId: String, event: GroupMessageEvent) {
        try {
            val enabled = CommandRepositories.groupSettings.fullForwarding(groupId, plugin.fullAmount)
            if (!enabled || !plugin.chatFormat.postChat) return

            var senderName = "unknown"
            val contact: Contact? = event.sender
            if (contact?.username != null) senderName = contact.username!!

            val baseMessage = event as? BaseMessageEvent<*>
            val metadata: JSONObject? = baseMessage?.metadata
            val attachments: JSONArray? = metadata?.getJSONArray("attachments")

            val parts = ArrayList<String>()
            val rawContent = event.rawMessage?.content
            val trimmed = rawContent?.trim() ?: ""

            var hasImage = false
            if (attachments != null) {
                for (i in 0 until attachments.size) {
                    val attachment = attachments.getJSONObject(i) ?: continue
                    val contentType = attachment.getString("content_type") ?: ""
                    if (contentType.startsWith("image/")) {
                        hasImage = true
                        break
                    }
                }
            }
            if (!hasImage && trimmed.isNotEmpty()) parts.add(trimmed)
            if (attachments != null) {
                for (i in 0 until attachments.size) {
                    val attachment = attachments.getJSONObject(i) ?: continue
                    val contentType = attachment.getString("content_type") ?: ""
                    when {
                        "voice" == contentType -> {
                            val asr = attachment.getString("asr_refer_text")
                            parts.add(if (!asr.isNullOrBlank()) "[语音] [${asr.trim()}]" else "[语音]")
                        }
                        contentType.startsWith("image/") -> parts.add("[图片]")
                        "image/gif" == contentType -> parts.add("[表情包]")
                        contentType.startsWith("video/") -> parts.add("[视频]")
                        else -> parts.add("[文件: ${attachment.getString("filename") ?: "文件"}]")
                    }
                }
            }
            if (parts.isEmpty()) return

            var message = parts.joinToString(" ")
            // @ 片段还原为昵称
            val mentions: Array<User>? = event.rawMessage?.mentions
            if (mentions != null) {
                for (member in mentions) {
                    val memberId = member.id ?: continue
                    val memberName = member.username ?: continue
                    message = message.replace("<@!$memberId>", "@$memberName")
                        .replace("<@$memberId>", "@$memberName")
                        .replace("<$memberId>", "@$memberName")
                        .replace(memberId, "@$memberName")
                }
            }
            plugin.broadcastMessage(plugin.formatGroupMessage(senderName, plugin.auditText(message)))
        } catch (_: Throwable) {
        }
    }
}
